pkg/frontend/subscriptions_put_test.go

```diff
@@ -0,0 +1,471 @@
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/metrics/noop"
+	mock_database "github.com/Azure/ARO-RP/pkg/util/mocks/database"
+	utiltls "github.com/Azure/ARO-RP/pkg/util/tls"
+	"github.com/Azure/ARO-RP/test/util/listener"
+)
+
+func TestPutSubscription(t *testing.T) {
+	ctx := context.Background()
+
+	clientkey, clientcerts, err := utiltls.GenerateKeyAndCertificate("client", nil, nil, false, true)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	serverkey, servercerts, err := utiltls.GenerateKeyAndCertificate("server", nil, nil, false, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pool := x509.NewCertPool()
+	pool.AddCert(servercerts[0])
+
+	cli := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: pool,
+				Certificates: []tls.Certificate{
+					{
+						Certificate: [][]byte{clientcerts[0].Raw},
+						PrivateKey:  clientkey,
+					},
+				},
+			},
+		},
+	}
+
+	mockSubID := "00000000-0000-0000-0000-000000000000"
+
+	type test struct {
+		name           string
+		request        func(*api.Subscription)
+		mocks          func(*mock_database.MockSubscriptions)
+		wantStatusCode int
+		wantResponse   func() *api.Subscription
+		wantError      string
+	}
+
+	for _, tt := range []*test{
+		{
+			name: "add a new subscription - registered state",
+			request: func(sub *api.Subscription) {
+				sub.State = api.SubscriptionStateRegistered
+			},
+			mocks: func(subscriptions *mock_database.MockSubscriptions) {
+				subscriptions.EXPECT().
+					Get(gomock.Any(), mockSubID).
+					Return(nil, &cosmosdb.Error{StatusCode: http.StatusNotFound})
+
+				subDoc := &api.SubscriptionDocument{
+					ID: mockSubID,
+					Subscription: &api.Subscription{
+						State: api.SubscriptionStateRegistered,
+					},
+				}
+
+				subscriptions.EXPECT().
+					Create(gomock.Any(), subDoc).
+					Return(subDoc, nil)
+			},
+			wantStatusCode: http.StatusCreated,
+			wantResponse: func() *api.Subscription {
+				return &api.Subscription{
+					State: api.SubscriptionStateRegistered,
+				}
+			},
+		},
+		{
+			name: "add a new subscription - warned state",
+			request: func(sub *api.Subscription) {
+				sub.State = api.SubscriptionStateWarned
+			},
+			mocks: func(subscriptions *mock_database.MockSubscriptions) {
+				subscriptions.EXPECT().
+					Get(gomock.Any(), mockSubID).
+					Return(nil, &cosmosdb.Error{StatusCode: http.StatusNotFound})
+
+				subDoc := &api.SubscriptionDocument{
+					ID: mockSubID,
+					Subscription: &api.Subscription{
+						State: api.SubscriptionStateWarned,
+					},
+				}
+
+				subscriptions.EXPECT().
+					Create(gomock.Any(), subDoc).
+					Return(subDoc, nil)
+			},
+			wantStatusCode: http.StatusCreated,
+			wantResponse: func() *api.Subscription {
+				return &api.Subscription{
+					State: api.SubscriptionStateWarned,
+				}
+			},
+		},
+		{
```

Comment:
I think I'd be ok without the test blowup for all the states.  If we really think we need it, we should probably have a for loop over the states that are supposed to behave identically.  WDYT?

Reply:
@m1kola bump

Reply:
@jim-minter yes, some states behave in the same way. But taking into account cascading effect of some of the states (limit\block operations on clusters or even delete clusters), I think it's a good idea to have coverage for alll states to make sure thate we don't break this behaviour. 

But I agree that there are many almost identical test cases. I'll look into how to reduce the amount of test code here.

Reply:
@jim-minter I updated the test. Removed the `mock` setup function - this was the most verbose part: it now happends in the test body. The test cases now just contain initial state (object or an error on get from db) and the desired db state & response.

Not sure if it makes sense to try and reduce number of lines even further: I think it will be less readable.